src/pybind/mgr/dashboard/controllers/host.py

```diff
@@ -304,13 +307,22 @@ class Host(RESTController):
                  },
                  responses={200: LIST_HOST_SCHEMA})
     @RESTController.MethodMap(version=APIVersion(1, 2))
-    def list(self, sources=None, facts=False):
+    def list(self, sources=None, facts=False, offset: int = 0, 
+             limit: int = 5, search: str = '', sort: str = ''):
         hosts = get_hosts(sources)
+        params = ['hostname']
+        paginator = ListPaginator(int(offset), int(limit), sort, search, hosts,
+                                  searchable_params=params, sortable_params=params,
+                                  default_sort='+hostname')
+        hosts = [host for host in paginator.list()]
         orch = OrchClient.instance()
+        cherrypy.response.headers['X-Total-Count'] = paginator.get_count()
         if str_to_bool(facts):
             if orch.available():
                 if not orch.get_missing_features(['get_facts']):
-                    hosts_facts = orch.hosts.get_facts()
+                    hosts_facts = []
+                    for host in hosts:
+                        hosts_facts.append(orch.hosts.get_facts(host['hostname'])[0])
```

Comment:
it might not be better to put this in a loop. the `get_facts` endpoint is a bit slow. its good to test it somehow..

Reply:
This is the loop of paginated hosts, so you would expect something like calling 5 hosts, 10 hosts or if you are feeling naughty, 25 hosts max.

Reply:
Before we had get_facts without hostname specified which means it will list ALL facts for ALL hosts. So this will reduce the search space by paginating even though we now have a loop.

Reply:
```py
    # I'd put this structure outside the method (e.g.: as a decorator),
    # since all the data here is static and doesn't need the method scope
    # for anything
    @paginate(searchable_params=['hostname'], sortable_params=['hostname'], default_sort='+hostname')
    def list(self, sources=None, facts=False, offset: int = 0, limit: int = 5, search: str = '', sort: str = ''):
        orch = OrchClient.instance()

        add_facts = str_to_bool(facts) and orch.available() and not orch.get_missing_features(['get_facts'])

        # paginator... ¬¬
        cherrypy.response.headers['X-Total-Count'] = paginator.get_count()
        for host in paginator(get_hosts(sources), int(offset), int(limit), sort, search):
            host['service_instances'] = populate_service_instances(host['hostname'], host['services'])
            if add_facts:
                # I know this is not exactly how it worked before, but I hate this
                # idea of merging dictionary without any criteria
                host['facts'] = orch.hosts.get_facts(host['hostname'])[0]
```

Reply:
How is this better? We went from explicitly getting the paginator and using it as we've always done programming to a obscure decorator that who knows how it will be implemented and that you have to search to make sure it is doing what it does. Then you have to understand that a local variable is magically there! 

Tbh I completely disagree on these changes, it brings too many complexity by over abstracting.